tools/nnyeah/nnyeah/AssemblyComparator/ComparingVisitor.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.MaciOS.Nnyeah;
+using Mono.Cecil;
+
+#nullable enable
+
+namespace Microsoft.MaciOS.AssemblyComparator {
+	public class ComparingVisitor {
+		ModuleDefinition EarlierModule, LaterModule;
+		bool PublicOnly;
+
+		public ComparingVisitor (ModuleDefinition earlierModule, ModuleDefinition laterModule, bool publicOnly)
+		{
+			EarlierModule = earlierModule;
+			LaterModule = laterModule;
+			PublicOnly = publicOnly;
+		}
+
+		public void Visit ()
+		{
+			var earlierElements = ModuleElements.Import (EarlierModule, PublicOnly);
+			if (earlierElements is null)
+				throw new Exception (Errors.E0007);
+			var laterElements = ModuleElements.Import (LaterModule, PublicOnly);
+			if (laterElements is null)
+				throw new Exception (Errors.E0007);
+			var reworker = new TypeReworker (EarlierModule);
+			VisitTypes (reworker, earlierElements, laterElements);
+		}
+
+		void VisitTypes (TypeReworker reworker, ModuleElements earlier, ModuleElements later)
+		{
+			foreach (var typeName in earlier.Types.Keys) {
+				if (!later.Types.TryGetValue (typeName, out var laterElems)) {
+					TypeNotFound.Invoke (this, new TypeNotFoundEventArgs (typeName));
+					continue;
+				} else {
+					TypeFound.Invoke (this, new TypeFoundEventArgs (typeName, laterElems.DeclaringType.ToString ()));
+				}
+				if (!earlier.Types.TryGetValue (typeName, out var earlierElems)) {
+					throw new Exception (Errors.E0007);
+				}
+				VisitAllMembers (reworker, earlierElems, laterElems);
+			}
+		}
+
+		void VisitAllMembers (TypeReworker reworker, TypeElements earlier, TypeElements later)
+		{
+			VisitMembers (reworker, earlier.Methods, later.Methods);
+			VisitMembers (reworker, earlier.Fields, later.Fields);
+			VisitMembers (reworker, earlier.Events, later.Events);
+			VisitMembers (reworker, earlier.Properties, later.Properties);
+		}
+
+		void VisitMembers<T> (TypeReworker reworker,
+			List<TypeElement<T>> earlier, List<TypeElement<T>> later) where T : IMemberDefinition
+		{
+			foreach (var earlierElem in earlier) {
+				VisitMember (reworker, earlierElem, later);
+			}
+		}
+
+		void VisitMember<T> (TypeReworker reworker, TypeElement<T> elem,
+			List<TypeElement<T>> later) where T : IMemberDefinition
+		{
+			foreach (var late in later) {
```

Comment:
This is one of those rare cases where it will make a noticeable performance difference. It's funny because Sebastien used to give me grief for using LINQ for function/readability reasons and each time I had to point out where the surrounding performance (running multiple compilers) would overshadow the cost of using LINQ. In this case, the single limiting factor in this code is how fast it can search lists and in this case `FirstOrDefault` is nearly twice as slow as an inline `foreach`. Since we are searching on the order of a million times, it counts here. Conscious decision.